File: BlueMS/src/main/java/com/st/BlueMS/demos/PedometerFragment.java
/*
 * Copyright (c) 2017  STMicroelectronics – All rights reserved
 * The STMicroelectronics corporate logo is a trademark of STMicroelectronics
 *
 * Redistribution and use in source and binary forms, with or without modification,
 * are permitted provided that the following conditions are met:
 *
 * - Redistributions of source code must retain the above copyright notice, this list of conditions
 *   and the following disclaimer.
 *
 * - Redistributions in binary form must reproduce the above copyright notice, this list of
 *   conditions and the following disclaimer in the documentation and/or other materials provided
 *   with the distribution.
 *
 * - Neither the name nor trademarks of STMicroelectronics International N.V. nor any other
 *   STMicroelectronics company nor the names of its contributors may be used to endorse or
 *   promote products derived from this software without specific prior written permission.
 *
 * - All of the icons, pictures, logos and other images that are provided with the source code
 *   in a directory whose title begins with st_images may only be used for internal purposes and
 *   shall not be redistributed to any third party or modified in any way.
 *
 * - Any redistributions in binary form shall not include the capability to display any of the
 *   icons, pictures, logos and other images that are provided with the source code in a directory
 *   whose title begins with st_images.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND ANY EXPRESS OR
 * IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY
 * AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER
 * OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
 * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
 * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
 * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR
 * OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY
 * OF SUCH DAMAGE.
 */

package com.st.BlueMS.demos;


import android.animation.AnimatorInflater;
import android.animation.AnimatorSet;
import android.content.res.Resources;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.TextView;

import com.st.BlueMS.R;
import com.st.BlueSTSDK.Feature;
import com.st.BlueSTSDK.Features.FeaturePedometer;
import com.st.BlueSTSDK.Node;
import com.st.BlueSTSDK.gui.demos.DemoDescriptionAnnotation;
import com.st.BlueSTSDK.gui.demos.DemoFragment;

@DemoDescriptionAnnotation(name="Pedometer", requareAll = {FeaturePedometer.class},iconRes = R
        .drawable.pedometer_demo_icon)
public class PedometerFragment extends DemoFragment {


    private String mStepCountTextFormat;
    private String mStepFrequencyTextFormat;

    private FeaturePedometer mPedometer;

    private TextView mStepsCount;
    private TextView mStepsFrequency;

    private boolean mFlipPosition;
    private AnimatorSet mFlipImageLeft;
    private AnimatorSet mFlipImageRight;

    private void animatePedometerIcon(){
        if(mFlipImageRight.isRunning() || mFlipImageLeft.isRunning())
            return;

        if(mFlipPosition)
            mFlipImageLeft.start();
        else
            mFlipImageRight.start();
        mFlipPosition=!mFlipPosition;
    }

    private Feature.FeatureListener mPedometerListener = new Feature.FeatureListener() {

        @Override
        public void onUpdate(Feature f,final Feature.Sample sample) {
            final String stepCountStr = String.format(mStepCountTextFormat,
                    FeaturePedometer.getSteps(sample));
            int frequency =FeaturePedometer.getFrequency(sample);
            final String stepFreqStr = frequency>0 ?
                    String.format(mStepFrequencyTextFormat,frequency,
                            FeaturePedometer.FEATURE_UNIT[1]) : "";
            updateGui(new Runnable() {
                @Override
                public void run() {
                    mStepsCount.setText(stepCountStr);
                    mStepsFrequency.setText(stepFreqStr);
                    animatePedometerIcon();
                }
            });
        }
    };

    public PedometerFragment() {
        // Required empty public constructor
    }

    @Override
    protected void enableNeededNotification(@NonNull Node node) {
        mPedometer = node.getFeature(FeaturePedometer.class);
        if (mPedometer != null) {
            mPedometer.addFeatureListener(mPedometerListener);
            node.enableNotification(mPedometer);
            //read the current state for initialize the string data
            node.readFeature(mPedometer);
        }
    }

    @Override
    protected void disableNeedNotification(@NonNull Node node) {
        if (mPedometer != null) {
            mPedometer.removeFeatureListener(mPedometerListener);
            node.disableNotification(mPedometer);
        }
    }


    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {
        // Inflate the layout for this fragment
        View root = inflater.inflate(R.layout.fragment_pedometer_feature, container, false);

        mStepsCount = (TextView) root.findViewById(R.id.stepCount);
        mStepsFrequency = (TextView) root.findViewById(R.id.stepFrequency);
        ImageView pedometerImage = (ImageView) root.findViewById(R.id.stepImage);
        pedometerImage.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                if(mPedometer!=null){
                    Node node = getNode();
                    if(node!=null)
                        node.readFeature(mPedometer);
                }
            }
        });

        mFlipImageLeft = (AnimatorSet) AnimatorInflater.loadAnimator(getActivity(),
                R.animator.flip_image_right);
        mFlipImageLeft.setTarget(pedometerImage);

        mFlipImageRight = (AnimatorSet) AnimatorInflater.loadAnimator(getActivity(),
                R.animator.flip_image_left);
        mFlipImageRight.setTarget(pedometerImage);

        final Resources res = getResources();
        mStepCountTextFormat = res.getString(R.string.stepCounterStringFormat);
        mStepFrequencyTextFormat = res.getString(R.string.stepFrequencyStringFormat);

        return root;
    }

}

File: st_pedometer/src/main/java/com/st/pedometer/PedometerFragment.kt
/*
 * Copyright (c) 2022(-0001) STMicroelectronics.
 * All rights reserved.
 * This software is licensed under terms that can be found in the LICENSE file in
 * the root directory of this software component.
 * If no LICENSE file comes with this software, it is provided AS-IS.
 */
package com.st.pedometer

import android.animation.AnimatorInflater
import android.animation.AnimatorSet
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.TextView
import androidx.fragment.app.Fragment
import androidx.fragment.app.viewModels
import androidx.lifecycle.Lifecycle
import androidx.lifecycle.lifecycleScope
import androidx.lifecycle.repeatOnLifecycle
import com.st.blue_sdk.features.pedometer.PedometerInfo
import com.st.core.ARG_NODE_ID
import com.st.pedometer.databinding.PedometerFragmentBinding
import dagger.hilt.android.AndroidEntryPoint
import kotlinx.coroutines.launch

@AndroidEntryPoint
class PedometerFragment : Fragment() {

    private val viewModel: PedometerViewModel by viewModels()
    private lateinit var binding: PedometerFragmentBinding
    private lateinit var nodeId: String

    private lateinit var mStepsCount: TextView
    private lateinit var mStepsFrequency: TextView

    private var mFlipPosition = false
    private lateinit var mFlipImageLeft: AnimatorSet
    private lateinit var mFlipImageRight: AnimatorSet

    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View {
        nodeId = arguments?.getString(ARG_NODE_ID)
            ?: throw IllegalArgumentException("Missing string $ARG_NODE_ID arguments")

        binding = PedometerFragmentBinding.inflate(inflater, container, false)

        mStepsCount = binding.stepCount
        mStepsFrequency = binding.stepFrequency
        val pedometerImage = binding.stepImage

        pedometerImage.setOnClickListener {
            viewModel.readFeature(nodeId)
        }

        mFlipImageLeft = AnimatorInflater.loadAnimator(
            activity,
            R.animator.flip_image_right
        ) as AnimatorSet
        mFlipImageLeft.setTarget(pedometerImage)

        mFlipImageRight = AnimatorInflater.loadAnimator(
            activity,
            R.animator.flip_image_left
        ) as AnimatorSet
        mFlipImageRight.setTarget(pedometerImage)

        return binding.root
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)

        viewLifecycleOwner.lifecycleScope.launch {
            viewLifecycleOwner.repeatOnLifecycle(Lifecycle.State.STARTED) {
                viewModel.stepData.collect {
                    updateGui(it)
                }
            }
        }
    }

    private fun updateGui(it: PedometerInfo) {
        val steps = it.steps.value
        val frequency = it.frequency.value

        val stepCountStr = String.format("%1d steps", steps)
        val stepFreqStr = String.format(" %1d %2s", frequency, it.frequency.unit)

        //Set Strings
        mStepsCount.text = stepCountStr
        mStepsFrequency.text = stepFreqStr

        //Animate the Icon
        if (mFlipImageRight.isRunning || mFlipImageLeft.isRunning) {
            return
        }

        if (mFlipPosition) {
            mFlipImageLeft.start()
        } else {
            mFlipImageRight.start()
        }
        mFlipPosition = !mFlipPosition
    }

    override fun onResume() {
        super.onResume()
        viewModel.startDemo(nodeId = nodeId)
    }

    override fun onPause() {
        super.onPause()
        viewModel.stopDemo(nodeId = nodeId)
    }
}
